spring-batch-infrastructure/src/main/java/org/springframework/batch/item/data/builder/RepositoryItemReaderBuilder.java

```diff
@@ -0,0 +1,214 @@
+	 * {@link #repository(PagingAndSortingRepository)}.
+	 *
+	 * @param repositoryReference of the used to get a repository and type-safe method for
+	 * use by the reader.
+	 * @return The current instance of the builder.
+	 * @see RepositoryItemReader#setMethodName(String)
+	 * @see RepositoryItemReader#setRepository(PagingAndSortingRepository)
+	 *
+	 */
+	public RepositoryItemReaderBuilder<T> repository(RepositoryMethodReference repositoryReference) {
+		Assert.notNull(repositoryReference, "repositoryReference must not be null.");
+		this.methodName = repositoryReference.getMethodName();
+		this.repository = repositoryReference.getRepository();
+
+		return this;
+	}
+
+	/**
+	 * Builds the {@link RepositoryItemReader}.
+	 *
+	 * @return a {@link RepositoryItemReader}
+	 */
+	public RepositoryItemReader<T> build() {
+		Assert.notNull(this.sorts, "sorts map is required.");
+		Assert.notNull(this.repository, "repository is required.");
+		Assert.hasText(this.methodName, "methodName is required.");
+		if (this.saveState) {
+			Assert.state(StringUtils.hasText(this.name), "A name is required when saveState is set to true.");
+		}
+
+		RepositoryItemReader<T> reader = new RepositoryItemReader<>();
+		reader.setArguments(this.arguments);
+		reader.setRepository(this.repository);
+		reader.setMethodName(this.methodName);
+		reader.setPageSize(this.pageSize);
+		reader.setCurrentItemCount(this.currentItemCount);
+		reader.setMaxItemCount(this.maxItemCount);
+		reader.setSaveState(this.saveState);
+		reader.setSort(this.sorts);
+		reader.setName(this.name);
+		return reader;
+	}
+
+	/**
+	 * Establishes a proxy that will capture a the Repository and the associated
+	 * methodName that will be used by the reader.
+	 * @param <T> The type of repository that will be used by the reader.
+	 */
+	public static class RepositoryMethodReference<T> {
+		private RepositoryMethodIterceptor repositoryInvocationHandler;
+
+		private PagingAndSortingRepository<?, ?> repository;
+
+		public RepositoryMethodReference(PagingAndSortingRepository<?, ?> repository) {
+			this.repository = repository;
+			this.repositoryInvocationHandler = new RepositoryMethodIterceptor();
+		}
+
+		// T is a proxy of the object passed in in the constructor
+		public T methodIs() {
+
+			Enhancer enhancer = new Enhancer();
+			enhancer.setSuperclass(this.repository.getClass());
+			enhancer.setCallback(this.repositoryInvocationHandler);
+			return (T) enhancer.create();
+		}
+
+		public PagingAndSortingRepository<?, ?> getRepository() {
+			return this.repository;
+		}
+
+		public String getMethodName() {
+			return this.repositoryInvocationHandler.getMethodName();
+		}
+
+	}
+
+	private static class RepositoryMethodIterceptor implements MethodInterceptor {
+		private String methodName;
+
+		@Override
+		public Object intercept(Object o, Method method, Object[] objects, MethodProxy methodProxy) throws Throwable {
+			this.methodName = method.getName();
```

Comment:
Can't we save off the arguments here as well so they don't even need to configure the arguments call?